net/xia/ppal_ether/main.c

```diff
@@ -191,10 +415,50 @@ static struct ether_interface *eint_init(struct net_device *dev)
 	return eint;
 }
 
+/* Caller must hold RTNL lock, and makes sure that nobody adds entries
+ * in eint->list_interface_common_addr while it's running.
+ */
+static void free_neighs_by_interface(struct ether_interface *eint)
+{
+	struct net_device *dev;
+	struct net *net;
+	struct xip_ppal_ctx *ctx;
+	struct fib_xid_table *xtbl;
+	struct fib_xid_ether_main *mether, *temp;
+
+	ASSERT_RTNL();
+
+	dev = eint->dev;
+	net = dev_net(dev);
+	ctx = xip_find_my_ppal_ctx_vxt(net, ether_vxt);
+	xtbl = ctx->xpc_xtbl;
+
+	list_for_each_entry_safe(mether, temp,
+				 &eint->list_interface_common_addr,
+				 interface_common_addr){
+		u8 xid[XIA_XID_MAX];
+		struct fib_xid *fxid;
+		u32 bucket;
+
+		memmove(xid, mether->xem_common.fx_xid, XIA_XID_MAX);
+
+		/* We don't lock eint->interface_lock to avoid deadlock. */
+		fxid = ether_rt_iops->fxid_find_lock(&bucket, xtbl, xid);
+		if (fxid) {
+			BUG_ON(fxid->fx_table_id != XRTABLE_MAIN_INDEX);
```

Comment:
Add after this `BUG_ON()` another test: `BUG_ON(&mether->xem_common != fxid);`.

Reply:
Is it not redundant as we are already fetching the fxid using the mether's xid along with a lock over the table?

Reply:
The suggested `BUG_ON()` is there to catch weird bugs that can be eventually introduced like this one: there are two fib XID entries with the **same** XID, and `fxid_find_lock()` returns the other entry instead of the one we are looking for. Of course, we don't want to have these bugs in our code, but we are humans and will eventually add bugs. We insert these asserts throughout the code to make it easier to catch these, otherwise, hard-to-catch bugs.